File: client/javascript/src/prsim/datahouse.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {spi} from "@/macro/spi";
import {mpi} from "@/macro/mpi";
import type {Context} from "@/prsim/context";
import {Page, Paging} from "@/kinds/page";
import {Doc} from "@/kinds/doc";
import {Status} from "@/cause";
import {index} from "@/macro";

@spi("mesh")
export abstract class DataHouse {

    // Writes
    @mpi("mesh.dh.writes")
    writes(ctx: Context, @index(0, 'docs') docs: Doc): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    // Write
    @mpi("mesh.dh.write")
    write(ctx: Context, @index(0, 'doc') doc: Doc): Promise<void> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    // Read
    @mpi("mesh.dh.read", [Page, Object])
    read(ctx: Context, @index(0, 'index') index: Paging): Promise<Page<any>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    // Indies
    @mpi("mesh.dh.indies", [Page, Object])
    indies(ctx: Context, @index(0, 'index') index: Paging): Promise<Page<any>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

    // Tables
    @mpi("mesh.dh.tables", [Page, Object])
    tables(ctx: Context, @index(0, 'index') index: Paging): Promise<Page<any>> {
        return Promise.reject(Status.URN_NOT_PERMIT)
    }

}

File: client/golang/prsim/datahouse.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package prsim

import (
	"context"
	"github.com/be-io/mesh/client/golang/types"
)

var IDataHouse = (*DataHouse)(nil)

// DataHouse spi
// @MPI(macro.MeshMPI)
// @SPI(macro.MeshSPI)
type DataHouse interface {

	// Writes
	// @MPI("mesh.dh.writes")
	Writes(ctx context.Context, docs []*types.Document) error

	// Write
	// @MPI("mesh.dh.write")
	Write(ctx context.Context, doc *types.Document) error

	// Read
	// @MPI("mesh.dh.read")
	Read(ctx context.Context, index *types.Paging) (*types.Page[any], error)

	// Indies
	// @MPI("mesh.dh.indies")
	Indies(ctx context.Context, index *types.Paging) (*types.Page[any], error)

	// Tables
	// @MPI("mesh.dh.tables")
	Tables(ctx context.Context, index *types.Paging) (*types.Page[any], error)
}
